### PR TITLE
Reduce C code

### PR DESCRIPTION
**Pull request by qmuntal**
This PR moves almost all C logic to Go. The benefits of this change are:
- Ease of debugging. It is more easier to debug Go code than C, and almost impossible to debug Go and C at the same time.
- Better error reporting. When moving errors from C to Go there is always some information lost, as it is done mainly using error codes.
- Less complexity. With this PR we no longer follow the bridge strategy nor use `DEFINEFUNC_FALLBACK`. It was confusing even for me: what are we bridging? How do we bridge a function defined in OpenSSL 1.1.1 but undefined in 1.0.2 and 3.0? etc. 
We now support different OpenSSL versions using good old `if vMajor == 3 {...} else {...}`

There are a couple of things that can remain in C:
- `go_openssl_load_functions`: This function is expanding the `FOR_ALL_OPENSSL_FUNCTIONS` macro, which defines necessary C functions and function pointers.
- `go_openssl_thread_setup`: This function implements logic to ensure OpenSSL 1.0.2 is thread-safe. Doing it in Go would case a big latency, as it is done using function callbacks and  worth nothing that these are slow in cgo.

This PR can be reviewed commit by commit.

### CODE REVIEW

File: openssl/openssl.go [hunk 1]
```diff
@@ -24,22 +27,112 @@ var (
...
+
+func errUnsuportedVersion() error {
+	return errors.New("openssl: OpenSSL version: " + strconv.Itoa(vMajor) + "." + strconv.Itoa(vMinor))
+}
+
 // Init loads and initializes OpenSSL.
 // It must be called before any other OpenSSL call.
+//
+// Only the first call to Init is effective,
+// subsecuent calls will return the same error result as the one from the first call.
```
Comment (chsienki) [inline openssl/openssl.go:61]:
```suggestion
// subsequent calls will return the same error result as the one from the first call.
```

**Review by chsienki — APPROVED** [review]

**Review by dagood — APPROVED** [review]

File: openssl/sha.go [hunk 2]
```diff
@@ -489,47 +499,36 @@ func (h *sha512Hash) UnmarshalBinary(b []byte) error {
...
+	return append(b,
+		byte(x>>56),
+		byte(x>>48),
+		byte(x>>40),
+		byte(x>>32),
+		byte(x>>24),
+		byte(x>>16),
+		byte(x>>8),
+		byte(x),
+	)
```
[the remark this answers is not in the document]
Comment (qmuntal) [reply openssl/sha.go:513]: This is just simplifying `appendUint64` so its easier to understand. As a side effect `appendUint32` is now faster, but it was already very fast so no one will notice.

```
name             old speed      new speed      delta
appendUint32-12  5.17GB/s ± 5%  8.19GB/s ±10%  +58.50%  (p=0.000 n=10+9)
appendUint64-12   340MB/s ±20%   358MB/s ±10%     ~     (p=0.315 n=10+10)

name             old B/op   new B/op   delta
appendUint32-12     0.00B          0.00B          ~     (all equal)
appendUint64-12     16.0B ± 0%     16.0B ± 0%     ~     (all equal)

name             old allocs/op  new allocs/op  delta
appendUint32-12      0.00           0.00          ~     (all equal)
appendUint64-12      1.00 ± 0%      1.00 ± 0%     ~     (all equal)
```

File: openssl/rsa.go [hunk 3]
```diff
@@ -206,3 +201,103 @@ func VerifyRSAPKCS1v15(pub *PublicKeyRSA, h crypto.Hash, hashed, sig []byte) err
...
+	if vMajor == 1 && vMinor == 0 {
+		r := (*rsa_st_1_0_2)(unsafe.Pointer(key))
+		//r.d and d will be nil for public keys.
+		if (r.n == nil && n == nil) ||
+			(r.e == nil && e == nil) {
+			return false
+		}
+		bnSet(&r.n, n)
+		bnSet(&r.e, e)
+		bnSet(&r.d, d)
```
[the remark this answers is not in the document]
Comment (qmuntal) [reply openssl/rsa.go:236]: Yep, calling a couple of `memset` more won't be noticeable, the bottleneck here is cgo context switch.

In fact, we are always calling `rsaSetKey` and friends with an uninitialized RSA key, so there is nothing to free. I added those `BN_clear_free` to be in the safe side.

File: openssl/openssl.go [hunk 4]
```diff
@@ -24,22 +27,108 @@ var (
...
+		if v1_sentinel != nil {
+			vMajor = 1
+			if v1_0_sentinel != nil {
+				vMinor = 0
+			} else {
+				vMinor = 1
+			}
+		} else {
+			vMajor = 3
+			vMinor = 0
```
[the remark this answers is not in the document]
Comment (qmuntal) [reply openssl/openssl.go:92]: This last comment made me realize that inferring the OpenSSL version by checking the presence of certain symbols is prone to error, as distros can patch OpenSSL to define symbols removed from upstream OpenSSL for the sake of compatibility.

I will submit another PR switching from symbol checking to using something like [OpenSSL_version_num](https://beta.openssl.org/docs/manmaster/man3/OpenSSL_version_num.html), which can be called without initializing OpenSSL.

**Pull request merged**